pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java

```diff
@@ -2826,13 +2834,16 @@ public void readStartupMessages() throws IOException, SQLException {
         case 'K':
           // BackendKeyData
           int msgLen = pgStream.receiveInteger4();
-          if (msgLen != 12) {
-            throw new PSQLException(GT.tr("Protocol error.  Session setup failed."),
-                PSQLState.PROTOCOL_VIOLATION);
-          }
-
           int pid = pgStream.receiveInteger4();
-          int ckey = pgStream.receiveInteger4();
+          int keyLen = msgLen - 8;
+          byte[] ckey = new byte[keyLen];
+          if ( getProtocolMajorVersion() == 3 && getProtocolMinorVersion() == 0 ) {
+            if (msgLen != 12) {
+              throw new PSQLException(GT.tr("Protocol error.  Session setup failed."),
+                  PSQLState.PROTOCOL_VIOLATION);
```

Comment:
Please include the failure detains into the message. Something like "cancel key should be 4 bytes only for protocol version 3.0. Got .. from the backend" would probably be easier to understand for those who get the exception in their systems.